aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/customize/s3/S3ExpressDecorator.kt

```diff
@@ -0,0 +1,313 @@
+                            self
+                        }
+                        """,
+                        *codegenScope,
+                    )
+                }
+
+                else -> emptySection
+            }
+        }
+}
+
+class S3ExpressFluentClientCustomization(
+    codegenContext: ClientCodegenContext,
+) : FluentClientCustomization() {
+    private val runtimeConfig = codegenContext.runtimeConfig
+    private val codegenScope =
+        arrayOf(
+            *preludeScope,
+            "S3ExpressRuntimePlugin" to s3ExpressModule(runtimeConfig).resolve("runtime_plugin::S3ExpressRuntimePlugin"),
+        )
+
+    override fun section(section: FluentClientSection): Writable =
+        writable {
+            when (section) {
+                is FluentClientSection.AdditionalBaseClientPlugins -> {
+                    rustTemplate(
+                        """
+                        ${section.plugins} = ${section.plugins}.with_client_plugin(#{S3ExpressRuntimePlugin}::new(${section.config}.config.clone()));
+                        """,
+                        *codegenScope,
+                    )
+                }
+
+                else -> emptySection
+            }
+        }
+}
+
+class S3ExpressRequestChecksumCustomization(
+    codegenContext: ClientCodegenContext,
+    private val operationShape: OperationShape,
+) : OperationCustomization() {
+    private val runtimeConfig = codegenContext.runtimeConfig
+
+    private val codegenScope =
+        arrayOf(
+            *preludeScope,
+            "ChecksumAlgorithm" to RuntimeType.smithyChecksums(runtimeConfig).resolve("ChecksumAlgorithm"),
+            "ConfigBag" to RuntimeType.configBag(runtimeConfig),
+            "DefaultRequestChecksumOverride" to
+                runtimeConfig.awsInlineableHttpRequestChecksum()
+                    .resolve("DefaultRequestChecksumOverride"),
+            "Document" to RuntimeType.smithyTypes(runtimeConfig).resolve("Document"),
+            "for_s3_express" to s3ExpressModule(runtimeConfig).resolve("utils::for_s3_express"),
+        )
+
+    override fun section(section: OperationSection): Writable =
+        writable {
+            when (section) {
+                is OperationSection.AdditionalRuntimePluginConfig -> {
+                    rustTemplate(
+                        """
+                        ${section.newLayerName}.store_put(#{DefaultRequestChecksumOverride}::new(
+                            |original: #{Option}<#{ChecksumAlgorithm}>,
+                            cfg: &#{ConfigBag}| {
+                                // S3 does not have the `ChecksumAlgorithm::Md5`, therefore customers cannot set it
+                                // from outside.
+                                if original != #{Some}(#{ChecksumAlgorithm}::Md5) {
+                                    return original;
+                                }
+
+                                if #{for_s3_express}(cfg) {
```

Comment:
Add a comment that S3 Express requires setting the checksum algorithm to CRC-32 when no other algorithm is provided.

Reply:
Addressed in 374ed2a